### PR TITLE
[Cache] Add __destruct() on TraceableAdapter to fix fatal error

### PR DESCRIPTION
**Pull request by andrerom**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | N/A?
| License       | MIT

Fixes:
```
PHP Fatal error:  Uncaught Error: Call to undefined method Symfony\Component\Cache\Adapter\TraceableTagAwareAdapter::__destruct() in /usr/sf/var/cache/dev/ContainerOjtvevg/TagAwareAdapter_5b4740c.php:128
Stack trace:
#0 [internal function]: TagAwareAdapter_5b4740c->__destruct()
```

When setting a TagAwareAdapter as cache service.
In my case it was a custom TagAwareAdapter _(which does not decorate other cache services)_, but I think the issue is probably also there if you use the default TagAwareAdapter and set it as cache service.

### CODE REVIEW

**Comment by andrerom** [comment]
```php
class TagAwareAdapter_5b4740c extends \Symfony\Component\Cache\Adapter\TagAwareAdapter implements \ProxyManager\Proxy\VirtualProxyInterface
(...)
    /**
     * {@inheritDoc}
     */
    public function __destruct()
    {
        $this->initializer5b4740c || $this->valueHolder5b4740c->__destruct();
    }
```

**Comment by andrerom** [comment]
config:
```yml
services:
    cache.filesystem:
        class: EzSystems\SymfonyTools\Incubator\Cache\TagAware\FilesystemTagAwareAdapter
        parent: cache.adapter.filesystem
        tags:
            - name: cache.pool
              clearer: cache.app_clearer
              namespace: '%cache_namespace%'
```

**Comment by nicolas-grekas** [comment]
That's a bug in @Ocramius/proxy-manager: it should not generate that call to destruct, isn't it?

File: src/Symfony/Component/Cache/Adapter/TraceableAdapter.php [hunk 1]
```diff
@@ -220,6 +220,11 @@ protected function start($name)
 
         return $event;
     }
+
+    public function __destruct()
+    {
+        unset($this->pool, $this->calls);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/TraceableAdapter.php]: I'd suggest removing these and replacing by `// no-op`

**Comment by nicolas-grekas** [comment]
can you try updating proxy-manager? maybe the bug is fixed already?

**Comment by andrerom** [comment]
> That's a bug in @Ocramius/proxy-manager: it should not generate that call to destruct, isn't it?

Could be, could also be bug in how framworkbundle wires up these classes, or something on our side on that. As in looks strange that it extends TagAwareAdapter.

So might be this is unneeded.

> can you try updating proxy-manager? maybe the bug is fixed already?

checking that and a few other things

**Comment by andrerom** [comment]
Updating Proxy Manager didn't change anything.

However fixing the service that is marked as lazy here fixed this:
```diff
 services:
     ezpublish.cache_pool:
-        class: Symfony\Component\Cache\Adapter\TagAwareAdapter
+        # As we support custom TagAware services, we set class as interface here so lazy class is "correct"
+        class: Symfony\Component\Cache\Adapter\TagAwareAdapterInterface
         lazy: true
         factory: ["@ezpublish.cache_pool.factory", getCachePool]
         arguments: ["@ezpublish.config.resolver"]
```

Thus closing this as invalid.

**Pull request closed without merging**